File: Ruby/0002.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
    dummy_head = ListNode.new(0) # Create a dummy head to simplify the code
    current = dummy_head         # Pointer to construct the new linked list
    carry = 0                   # Initialize carry to 0
  
    # Traverse both linked lists
    while l1 || l2 || carry != 0
      # Get the current values from the nodes, if present
      val1 = l1 ? l1.val : 0
      val2 = l2 ? l2.val : 0
  
      # Calculate the sum of current digits and the carry
      sum = val1 + val2 + carry
  
      # Update carry for the next digit
      carry = sum / 10
  
      # Create a new node with the digit part of the sum
      current.next = ListNode.new(sum % 10)
      current = current.next
  
      # Move to the next nodes in the input lists
      l1 = l1 ? l1.next : nil
      l2 = l2 ? l2.next : nil
    end
  
    # Return the next node of the dummy head, which is the start of the result list
    dummy_head.next
  end

File: C++/0002.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        // Initialize a dummy node to help with list creation
        ListNode dummy;
        ListNode* current = &dummy;
        int carry = 0;
        
        // Traverse both linked lists
        while (l1 != nullptr || l2 != nullptr || carry != 0) {
            // Get the values from the current nodes, or 0 if the node is nullptr
            int val1 = (l1 != nullptr) ? l1->val : 0;
            int val2 = (l2 != nullptr) ? l2->val : 0;
            
            // Calculate the sum and carry
            int sum = val1 + val2 + carry;
            carry = sum / 10;
            sum = sum % 10;
            
            // Create a new node with the computed sum value
            current->next = new ListNode(sum);
            current = current->next;
            
            // Move to the next nodes in the lists
            if (l1 != nullptr) l1 = l1->next;
            if (l2 != nullptr) l2 = l2->next;
        }
        
        // Return the next node after the dummy node, which is the start of the result list
        return dummy.next;
    }
};
